badge_server/main.py

```diff
@@ -75,519 +88,235 @@ def _get_pair_status_for_packages(pkg_sets):
+            result_dict[pyver]['status'] = res.status.value
+            result_dict[pyver]['details'][other_package_name] = details
+
+    return result_dict
+
+
+def _get_dependency_dict(package_name) -> dict:
+    """Return the dict which contains the self outdated dependencies
+    status and details.
+
+    Args:
+        package_name: the name of the package to get outdated dependencies for.
+
+    Returns:
+        A dict containing the outdated dependency status and details for any
+        outdated dependencies. The dict will be formatted like the following:
+
+        {
+            'status': 'UP_TO_DATE', 'details': {}
+        }
+    """
+    result_dict = badge_utils._build_default_result(
+        status='UP_TO_DATE', include_pyversion=False, details={})
+
+    outdated_deps = badge_utils.highlighter.check_package(package_name)
+    _deps_list = badge_utils.finder.get_deprecated_dep(package_name)[1]
+    deprecated_deps = ', '.join(_deps_list)
+
+    outdated_depencdency_name_to_details = {}
+    max_level = badge_utils.priority_level.UP_TO_DATE
+    for dep in outdated_deps:
+        dep_detail = {}
+        level = dep.priority.level
+        if level.value > max_level.value:
+            max_level = level
+        dep_detail['installed_version'] = dep.installed_version
+        dep_detail['latest_version'] = dep.latest_version
+        dep_detail['priority'] = dep.priority.level.name
+        dep_detail['detail'] = dep.priority.details
+        outdated_depencdency_name_to_details[dep.name] = dep_detail
+    result_dict['status'] = max_level.name
+    result_dict['details'] = outdated_depencdency_name_to_details
+    result_dict['deprecated_deps'] = deprecated_deps
+
+    return result_dict
+
+
+def _get_badge_status(
+        self_compat_res: dict,
+        google_compat_res: dict,
+        dependency_res: dict) -> str:
+    """Get the badge status.
+
+    The badge status will determine the right hand text and the color of
+    the badge.
+
+    Badge status to color mapping:
+    SUCCESS -> green
+    UNKNOWN -> purple
+    CHECK_WARNING -> red
+
+    See badge_utils.STATUS_COLOR_MAPPING.
+    """
+    dep_status = dependency_res['status']
+    dep_status = 'SUCCESS' if dep_status == 'UP_TO_DATE' else dep_status
 
-    response = flask.make_response(
-        pybadges.badge(
-            left_text=badge_name,
-            right_text=status,
-            right_color=color,
-            whole_link=details_link))
-    response.content_type = badge_utils.SVG_CONTENT_TYPE
-    response.headers['Cache-Control'] = 'no-cache'
-    response.add_etag()
+    _statuses = []
+    for pyver in ['py2', 'py3']:
+        _statuses.append(self_compat_res[pyver]['status'])
+        _statuses.append(google_compat_res[pyver]['status'])
+    _statuses.append(dep_status)
 
-    return response
+    if ['SUCCESS'] * len(_statuses) == _statuses:
+        return 'SUCCESS'
+    elif 'UNKNOWN' in _statuses:
```

Comment:
Is 'UNKNOWN' still possible?

Reply:
Yes.

In the current state of the badge server, 'UNKNOWN' will occur when a package is not whitelisted. I believe this is also in line with the PRD.

I know that in previous commits, I was handling for 'UNKNOWN' prior to this point in the code, but I have refactored the code such that the handling takes place here.